addons/snailmail/controller/snailmail_webhook.py

```diff
@@ -0,0 +1,83 @@
+import hashlib
+import hmac
+import json
+import logging
+
+from odoo import http
+from odoo.http import request
+from odoo.tools import consteq
+
+_logger = logging.getLogger(__name__)
+
+LETTER_STATUS_MAPPING = {
+    'undeliverable': 'Letter was unable to be delivered',
+    'delivered': 'Letter is delivered successfully'
+}
+
+
+class SnailmailWebhookController(http.Controller):
+    @http.route('/webhook/snailmail/<string:event_type>', type='jsonrpc', auth='public', save_session=False)
+    def snailmail_webhook(self, event_type):
+        """ Receive and process letter sent by IAP pingen webhook. """
+
+        event = request.get_json_data()
+        if not event or event_type not in LETTER_STATUS_MAPPING:
+            _logger.warning("[Snailmail] Webhook: Invalid request payload received")
+            raise request.not_found()
+
+        account_token = request.env['iap.account'].sudo().get('snailmail').account_token
+        hashed_token = hashlib.sha1(account_token.encode('utf-8')).hexdigest()
+        if not self._check_signature(hashed_token, event):
+            _logger.warning("[Snailmail] Webhook: IAP signature does not match")
+            raise request.not_found()
+
+        pingen_letter_id = event.get('letter_id')
+        letter_status = event.get('status')
+        letter = request.env['snailmail.letter'].sudo().search([('document_id', '=', pingen_letter_id)], limit=1)
+
+        if not letter:
+            _logger.warning("[Snailmail] Webhook: could not find a matching letter with letter_id = %s", pingen_letter_id)
+            raise request.not_found()
+
+        model_name, record_id = letter.reference.split(',')
+        if record := request.env[model_name].browse(int(record_id)):
+            letter_data = {'state': letter_status}
+            msg_id = record.sudo().message_post(
+                body=LETTER_STATUS_MAPPING.get(letter_status),
+                message_type='snailmail',
+                subtype_xmlid='mail.mt_note',
+                author_id=request.env.ref('base.partner_root').id
+            )
+            notification_vals = {
```

Comment:
Don't do that manually. Instead, use classic message post notification code (aka setting partner_ids=letter.partner_id.ids).